test/test_utils.h

```diff
@@ -159,23 +159,18 @@ bool compare_vectors(const std::vector<std::vector<T> *> &ref_vector,
 
     for (size_t i = 0; i < ref_vector.size(); i++) {
         for (size_t j = 0; j < ref_vector[i]->size(); j++) {
-            // We can't do the comparison directly because when the values are
-            // written and read from a file, they are converted to strings and
-            // back, which can cause some precision loss. So we convert them to
-            // strings and compare the strings.
-            std::stringstream aa;
-            aa << (*ref_vector[i])[j];
-            std::string a = aa.str();
-            std::stringstream bb;
-            bb << (*test_vector[i])[j];
-            std::string b = bb.str();
+            double test = (*test_vector[i])[j];
+            double ref = (*ref_vector[i])[j];
 
-            if (a != b) {
+            double diff = std::abs(test - ref);
+
+            if (diff > 1e-8) {
```

Comment:
If the if statement is triggered, the test will stop and print out all outputs. However, if we continue the run, it's unclear how many values will be printed out. Any ideas?

Reply:
I'm not sure what you mean by "it's unclear how many values will be printed out."
You mean like it would be useful to print all differences in the values not just the first to be found?

Reply:
For example, 
```C++
for (int i = 0; i < 10 i++){
    if (i == 4){
         ...
         return false;
     }
}
```
What I meant is that we couldn't see other failed values. I modified the code
```C++
int count + 
for (int i = 0; i < 10 i++){
    if (i == 4){
         ...
         count++;
     }
     if (count > 0){
           return false
     }
}
```
...but I realized when the failed values over 100 so it becomes irrelevant. We keep it as it is